File: src/c++/client/qos/PublishQos.cpp
/*------------------------------------------------------------------------------
Name:      PublishQos.cpp
Project:   xmlBlaster.org
Copyright: xmlBlaster.org, see xmlBlaster-LICENSE file
------------------------------------------------------------------------------*/

#include <client/qos/PublishQos.h>
#include <util/Global.h>

using namespace org::xmlBlaster::util;
using namespace org::xmlBlaster::util::qos;
using namespace std;

namespace org { namespace xmlBlaster { namespace client { namespace qos {

   PublishQos::PublishQos(Global& global)
      : ME("PublishQos"), global_(global), msgQosData_(global, "")
   {
      // deprecated
      //setLifeTime(global_.getProperty().getLongProperty("message.lifeTime", -1));
   }

   PublishQos::PublishQos(Global& global, const Destination& destination)
      : ME("PublishQos"), global_(global), msgQosData_(global, "")
   {
      addDestination(destination);
   }

   PublishQos::PublishQos(Global& global, bool persistent)
      : ME("PublishQos"), global_(global), msgQosData_(global, "")
   {
      setPersistent(persistent);
   }

   const MsgQosData& PublishQos::getData()
   {
      return msgQosData_;
   }

   PriorityEnum PublishQos::getPriority() const
   {
      return msgQosData_.getPriority();
   }

   /**
    * Set message priority value, PriorityEnum::NORM_PRIORITY (5) is default.
    * PriorityEnum::MIN_PRIORITY (0) is slowest
    * whereas PriorityEnum.MAX_PRIORITY (9) is highest priority.
    * @see <a href="http://www.xmlBlaster.org/xmlBlaster/doc/requirements/engine.qos.publish.priority.html">The engine.qos.publish.priority requirement</a>
    */
   void PublishQos::setPriority(PriorityEnum priority)
   {
      msgQosData_.setPriority(priority);
   }

   /**
    * Send message to subscriber even if the content is the same as the previous?
    * <br />
    * Default is that xmlBlaster does send messages to subscribed clients, even the content didn't change.
    */
   void PublishQos::setForceUpdate(bool force)
   {
      msgQosData_.setForceUpdate(force);
   }

   /**
    * Control message life cycle on message expiry. 
    * @param forceDestroy true Force message destroy on message expire<br />
    *        false On message expiry messages which are already in callback queues are delivered.
    */
   void PublishQos::setForceDestroy(bool forceDestroy)
   {
      msgQosData_.setForceDestroy(forceDestroy);
   }

   void PublishQos::setSubscribable(bool isSubcribeable)
   {
      msgQosData_.setSubscribable(isSubcribeable);
   }

   /**
    * Mark a message to be readonly.
    * <br />
    * Only the first publish() will be accepted, followers are denied.
    */
   void PublishQos::setReadonly(bool readonly)
   {
      msgQosData_.setReadonly(readonly);
   }

   /**
    * Mark a message to be volatile or not.
    * <br />
    * A non-volatile messages stays in memory as long as the server runs<br />
    * A volatile messages exists only during publish and processing it (doing the updates).<br />
    * Defaults to false.
    */
   void PublishQos::setVolatile(bool volatileFlag)
   {
      msgQosData_.setVolatile(volatileFlag);
   }

   /**
    * @see #isVolatile()
    */
   bool PublishQos::isVolatile()
   {
      return msgQosData_.isVolatile();
   }

   /**
    * Mark a message to be persistent.
    */
   void PublishQos::setPersistent(bool persistent)
   {
      msgQosData_.setPersistent(persistent);
   }

   /**
    * The message expires after given milliseconds (message is erased).<p />
    * Clients will get a notify about expiration.<br />
    * This value is calculated relative to the rcvTimestamp in the xmlBlaster server.<br />
    * Passing -1 milliseconds asks the server for unlimited livespan, which
    * the server may or may not grant.
    * @param lifeTime in milliseconds
    */
   void PublishQos::setLifeTime(long lifeTime)
   {
      msgQosData_.setLifeTime(lifeTime);
   }

   /**
    * Add a destination where to send the message.
    * <p />
    * Note you can invoke this multiple times to send to multiple destinations.
    * @param destination  The loginName of a receiver or some destination XPath query
    */
   void PublishQos::addDestination(const Destination& destination)
   {
      msgQosData_.addDestination(destination);
   }

   /**
    * Access sender name.
    * @return loginName of sender or null if not known
    */
   SessionNameRef PublishQos::getSender()
   {
      return msgQosData_.getSender();
   }

   /*
   void PublishQos::setSender(const SessionName& sender)
   {
      msgQosData_.setSender(sender);
   }
   */

   /**
    * @param state The state to return to the server.
    *   e.g. Contants.STATE_OK, see Constants::java
    */
   void PublishQos::setState(const string& state)
   {
      msgQosData_.setState(state);
   }

   string PublishQos::getState()
   {
      return msgQosData_.getState();
   }

   /**
    * @param stateInfo The state info attribute to return to the server.
    */
   void PublishQos::setStateInfo(const string& stateInfo)
   {
      msgQosData_.setStateInfo(stateInfo);
   }

   string PublishQos::getStateInfo()
   {
      return msgQosData_.getStateInfo();
   }

   /**
    * Administer/configure the message topic. 
    */
   void PublishQos::setTopicProperty(const TopicProperty& topicProperty)
   {
      msgQosData_.setTopicProperty(topicProperty);
   }

   void PublishQos::setClientProperties(const QosData::ClientPropertyMap& cm)
   {
      msgQosData_.setClientProperties(cm);
   }

   const QosData::ClientPropertyMap& PublishQos::getClientProperties() const
   {
      return msgQosData_.getClientProperties();
   }

   /**
    * Converts the data into a valid XML ASCII string.
    * @return An XML ASCII string
    */
   string PublishQos::toString()
   {
      return toXml();
   }

   /**
    * Converts the data into a valid XML ASCII string.
    * @return An XML ASCII string
    */
   string PublishQos::toXml()
   {
      return msgQosData_.toXml();
   }

}}}}

File: xmlBlaster/src/java/org/xmlBlaster/client/qos/PublishQos.java
/*------------------------------------------------------------------------------
Name:      PublishQos.java
Project:   xmlBlaster.org
Copyright: xmlBlaster.org, see xmlBlaster-LICENSE file
------------------------------------------------------------------------------*/
package org.xmlBlaster.client.qos;

import org.xmlBlaster.util.Global;
import org.xmlBlaster.util.XmlBlasterException;
import org.xmlBlaster.util.SessionName;
import org.xmlBlaster.util.enum.PriorityEnum;
import org.xmlBlaster.util.qos.MsgQosData;
import org.xmlBlaster.engine.helper.Destination;
import org.xmlBlaster.util.qos.TopicProperty;


/**
 * This class encapsulates the qos of a publish() message.
 * <p />
 * So you don't need to type the 'ugly' XML ASCII string by yourself.
 * After construction access the ASCII-XML string with the toXml() method.
 * <br />
 * A typical <b>publish</b> qos in Publish/Subcribe mode could look like this:<br />
 * <pre>
 *  &lt;qos>
 *     &lt;priority>5&lt;/priority>
 *     &lt;expiration lifeTime='60000' forceDestroy='false'/>
 *     &lt;isDurable />  &lt;!-- The message shall be recoverable if xmlBlaster crashes -->
 *     &lt;forceUpdate>true&lt;/forceUpdate>
 *     &lt;readonly />
 *  &lt;/qos>
 * </pre>
 * A typical <b>publish</b> qos in PtP mode could look like this:<br />
 * <pre>
 *  &lt;qos>
 *     &lt;destination queryType='EXACT' forceQueuing='true'>
 *        joe
 *     &lt;/destination>
 *     &lt;destination>
 *        /node/heron/client/Tim/-2
 *     &lt;/destination>
 *  &lt;/qos>
 * </pre>
 * <p />
 * see xmlBlaster/src/dtd/XmlQoS.xml
 * @see org.xmlBlaster.util.qos.MsgQosSaxFactory
 * @see <a href="http://www.xmlblaster.org/xmlBlaster/doc/requirements/interface.publish.html">publish interface</a>
 */
public final class PublishQos
{
   private String ME = "PublishQos";
   private final Global glob;
   private final MsgQosData msgQosData;

   /**
    * Default constructor for transient messages.
    */
   public PublishQos(Global glob) {
      this.glob = (glob==null) ? Global.instance() : glob;
      this.msgQosData = new MsgQosData(glob, glob.getMsgQosFactory()); 
      setLifeTime(org.xmlBlaster.util.Global.instance().getProperty().get("message.lifeTime", -1L)); // TODO: use local glob
   }

   /**
    * Default constructor for transient PtP messages.
    * <p />
    * To make the message persistent, use the durable() method
    * @param destination The object containing the destination address.<br />
    *        To add more destinations, us the addDestination() method.
    */
   public PublishQos(Global glob, Destination destination) {
      this(glob);
      addDestination(destination);
   }

   /**
    * @param isDurable true = store the message persistently
    */
   public PublishQos(Global glob, boolean durable) {
      this(glob);
      setDurable(durable);
   }

   public MsgQosData getData() {
      return this.msgQosData;
   }

   /**
    * Message priority.
    * @return priority 0 (=Lowest) - 9 (=Highest)
    */
   public PriorityEnum getPriority() {
      return msgQosData.getPriority();
   }

   /**
    * Set message priority value, PriorityEnum.NORM_PRIORITY (5) is default. 
    * PriorityEnum.MIN_PRIORITY (0) is slowest
    * whereas PriorityEnum.MAX_PRIORITY (9) is highest priority.
    * @see <a href="http://www.xmlBlaster.org/xmlBlaster/doc/requirements/engine.qos.publish.priority.html">The engine.qos.publish.priority requirement</a>
    */
   public void setPriority(PriorityEnum priority) {
      msgQosData.setPriority(priority);
   }

   /**
    * Send message to subscriber even if the content is the same as the previous?
    * <br />
    * Default is that xmlBlaster does send messages to subscribed clients, even the content didn't change.
    */
   public void setForceUpdate(boolean force) {
      msgQosData.setForceUpdate(force);
   }

   /**
    * Mark a message to be readonly.
    * <br />
    * Only the first publish() will be accepted, followers are denied.
    */
   public void setReadonly(boolean readonly) {
      msgQosData.setReadonly(readonly);
   }

   /**
    * Mark a message to be volatile or not.
    * <br />
    * A non-volatile messages stays in memory as long as the server runs<br />
    * A volatile messages exists only during publish and processing it (doing the updates).<br />
    * Defaults to false.
    * @deprecated Use setLifeTime(0L) and setForceDestroy(false) instead
    */
   public void setVolatile(boolean volatileFlag) {
      msgQosData.setVolatile(volatileFlag);
   }

   /**
    * @see #isVolatile()
    */
   public boolean isVolatile() {
      return msgQosData.isVolatile();
   }

   /**
    * Mark a message to be persistent.
    */
   public void setDurable(boolean durable) {
      msgQosData.setDurable(durable);
   }

   /**
    * The message expires after given milliseconds (message is erased).<p />
    * Clients will get a notify about expiration.<br />
    * This value is calculated relative to the rcvTimestamp in the xmlBlaster server.<br />
    * Passing -1 milliseconds asks the server for unlimited livespan, which
    * the server may or may not grant.
    * @param lifeTime in milliseconds
    */
   public void setLifeTime(long lifeTime) {
      msgQosData.setLifeTime(lifeTime);
   }

   /**
    * @param forceDestroy true Force message destroy on message expire<br />
    *        false On message expiry messages which are already in callback queues are delivered.
    */
   public void setForceDestroy(boolean forceDestroy) {
      msgQosData.setForceDestroy(forceDestroy);
   }

   /**
    * Add a destination where to send the message.
    * <p />
    * Note you can invoke this multiple times to send to multiple destinations.
    * @param destination  The loginName of a receiver or some destination XPath query
    */
   public void addDestination(Destination destination) {
      this.msgQosData.addDestination(destination);
   }

   /**
    * Access sender name.
    * @return loginName of sender or null if not known
    */
   public SessionName getSender() {
      return this.msgQosData.getSender();
   }

   /**
    * Access sender name.
    * @param loginName of sender
    */
   public void setSender(SessionName sender) {
      this.msgQosData.setSender(sender);
   }

   /**
    * @param state The state to return to the server.
    *   e.g. Contants.STATE_OK, see Constants.java
    */
   public void setState(String state) {
      this.msgQosData.setState(state);
   }

   public String getState() {
      return this.msgQosData.getState();
   }

   /**
    * @param stateInfo The state info attribute to return to the server.
    */
   public void setStateInfo(String stateInfo) {
      this.msgQosData.setStateInfo(stateInfo);
   }

   public String getStateInfo() {
      return this.msgQosData.getStateInfo();
   }

   /**
    * Administer/configure the message topic. 
    */
   public void setTopicProperty(TopicProperty topicProperty) {
      this.msgQosData.setTopicProperty(topicProperty);
   }

   /**
    * Converts the data into a valid XML ASCII string.
    * @return An XML ASCII string
    */
   public String toString() {
      return toXml();
   }

   /**
    * Converts the data into a valid XML ASCII string.
    * @return An XML ASCII string
    */
   public String toXml() {
      return msgQosData.toXml();
   }

   /**
    *  For testing invoke: java org.xmlBlaster.client.PublishQos
    */
   public static void main( String[] args ) throws XmlBlasterException
   {
      Global glob = new Global(args);
      {
         PublishQos qos =new PublishQos(new Global(args), new Destination(new SessionName(glob, "joe")));
         qos.addDestination(new Destination(new SessionName(glob, "Tim")));
         qos.setPriority(PriorityEnum.HIGH_PRIORITY);
         qos.setDurable(true);
         qos.setForceUpdate(true);
         qos.setReadonly(true);
         qos.setLifeTime(60000);
         System.out.println(qos.toXml());
      }
      {
         PublishQos qos =new PublishQos(null);
         System.out.println("Minimal '" + qos.toXml() + "'");
      }
   }
}
